auth/client/src/main/java/org/wildfly/security/credential/source/OAuth2CredentialSource.java

```diff
@@ -292,6 +300,105 @@ public Builder clientCredentials(String id, String secret) {
+            PasswordFactory factory = PasswordFactory.getInstance(algorithm);
+            MaskedPassword password = factory.generatePassword(spec).castAs(MaskedPassword.class);
+            ClearPasswordSpec clearSpec = factory.getKeySpec(password, ClearPasswordSpec.class);
+            String finalPassword = String.valueOf(clearSpec.getEncodedPassword());
+
+            configureAuthenticationHandler(parameters -> configureResourceOwnerCredentialsParameters(parameters, userName, finalPassword));
+            return this;
+        }
+
+        /**
+         * <p>Configure OAuth2 Masked Client Credentials Grant Type as defined by the OAuth2 specification.
+         *
+         * @param id the client id
+         * @param maskedSecret the masked password, as a string (must not be {@code null})
+         * @param algorithm the algorithm (can be {@code null}, default:"masked-MD5-DES")
+         * @param initialKeyMaterial the initial key material, as a string(can be {@code null}, default:"somearbitrarycrazystringthatdoesnotmatter")
+         * @param iterationCount the iteration count, as an integer (must not be less than 1)
+         * @param salt the salt, as a string (must not be {@code null})
+         * @param initializationVector the initialization vector, as a string (can be {@code null})
+         * @return this instance.
+         */
+        public Builder maskedClientCredentials(String id, String maskedSecret, String algorithm, String initialKeyMaterial, int iterationCount, String salt, String initializationVector) throws NoSuchAlgorithmException, InvalidKeySpecException {
+
+            Assert.assertNotNull(maskedSecret);
+            Assert.checkMinimumParameter("iterationCount", 1, iterationCount);
+            Assert.assertNotNull(salt);
+
+            byte[] finalMaskedSecretBytes = CodePointIterator.ofString(maskedSecret).base64Decode().drain();
+            if (algorithm == null) algorithm = MaskedPassword.ALGORITHM_MASKED_MD5_DES;
+            char[] finalInitialKeyMaterial = initialKeyMaterial == null ? "somearbitrarycrazystringthatdoesnotmatter".toCharArray() : initialKeyMaterial.toCharArray();
+            byte[] finalSalt = CodePointIterator.ofString(salt).asUtf8().drain();
+            byte[] finalInitializationVector = initializationVector == null ? null : CodePointIterator.ofString(initializationVector).base64Decode().drain();
+
+            MaskedPasswordSpec spec = new MaskedPasswordSpec(finalInitialKeyMaterial, iterationCount, finalSalt, finalMaskedSecretBytes, finalInitializationVector);
+            PasswordFactory factory = PasswordFactory.getInstance(algorithm);
+            MaskedPassword maskedPassword = factory.generatePassword(spec).castAs(MaskedPassword.class);
+            ClearPasswordSpec clearSpec = factory.getKeySpec(maskedPassword, ClearPasswordSpec.class);
+            String secret = String.valueOf(clearSpec.getEncodedPassword());
```

Comment:
Since the above lines are also performed in the above method, it might be good to create a private helper method in this class to avoid the duplication.

Reply:
Added